Python/specialize.c

```diff
@@ -1690,6 +1690,14 @@ _Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
                 goto success;
             }
             if (PyFloat_CheckExact(lhs)) {
+                if (adaptive->original_oparg == NB_INPLACE_ADD &&
+                    (_Py_OPCODE(instr[1]) == STORE_FAST ||
+                     _Py_OPCODE(instr[1]) == STORE_FAST__LOAD_FAST))
+                {
+                    *instr = _Py_MAKECODEUNIT(BINARY_OP_INPLACE_ADD_FLOAT,
```

Comment:
The line
```
for x, y in some_list:
    z = x + y
    ...
```

This will get specialized to INPLACE_ADD_FLOAT, then deopt because `z is not x`, and it will never make it to the other ADD_FLOAT specialization. Can we add a check for `locals[next_oparg] == lhs` at specialization time?

Reply:
I don’t think that’s true. Note that we check that `adaptive->original_oparg == NB_INPLACE_ADD` too before applying this specialization.

Unless `lhs` is rebound by a walrus or frame magic after being loaded (very unlikely, but protected by the additional runtime guard for the `INPLACE` variant), then `BINARY_OP(NB_INPLACE_ADD), STORE_FAST(...)` is sufficient to infer that this optimization is appropriate here.

(I guess we *could* pass a pointer to the locals array to this function, and use that to check whether the next store is indeed the same object. It would allow us to apply this same optimization to `lhs = lhs + rhs` too, but I’m not sure it’s worth the additional complexity… plus, we would still need the runtime check anyways.)

Reply:
Ah! I missed the `adaptive->original_oparg == NB_INPLACE_ADD`. Makes sense.